src/array/cpu/id_hash_map.cc

```diff
@@ -0,0 +1,195 @@
+/**
+ *  Copyright (c) latest by Contributors
+ * @file array/cpu/id_hash_map.cc
+ * @brief Class about id hash map
+ */
+
+#include "id_hash_map.h"
+
+#include <dgl/array.h>
+#include <dgl/runtime/device_api.h>
+#include <dgl/runtime/parallel_for.h>
+
+#include <cmath>
+#include <numeric>
+
+using namespace dgl::runtime;
+
+namespace {
+static constexpr int64_t kEmptyKey = -1;
+static constexpr int kGrainSize = 256;
+}  // namespace
+
+namespace dgl {
+namespace aten {
+
+template <typename IdType>
+IdType IdHashMap<IdType>::CompareAndSwap(
+    IdType* ptr, IdType old_val, IdType new_val) {
+#ifdef _MSC_VER
+  if (sizeof(IdType) == 4) {
+    return _InterlockedCompareExchange(
+        reinterpret_cast<LONG*>(ptr), new_val, old_val);
+  } else if (sizeof(IdType) == 8) {
+    return _InterlockedCompareExchange64(
+        reinterpret_cast<LONGLONG*>(ptr), new_val, old_val);
+  } else {
+    LOG(FATAL) << "ID can only be int32 or int64";
+  }
+#elif __GNUC__  // _MSC_VER
+  return __sync_val_compare_and_swap(ptr, old_val, new_val);
+#else           // _MSC_VER
+#error "CompareAndSwap is not supported on this platform."
+#endif  // _MSC_VER
+}
+
+template <typename IdType>
+IdHashMap<IdType>::IdHashMap() : hmap_(nullptr), mask_(0) {}
+
+template <typename IdType>
+IdArray IdHashMap<IdType>::Init(const IdArray ids) {
+  CHECK_EQ(ids.defined(), true);
+  const IdType* ids_data = ids.Ptr<IdType>();
+  const size_t num = static_cast<size_t>(ids->shape[0]);
+  CHECK_GT(num, 0);
+  size_t capcacity = 1;
+  capcacity = capcacity << static_cast<size_t>(1 + std::log2(num * 3));
+  mask_ = static_cast<IdType>(capcacity - 1);
+
+  DGLContext ctx = DGLContext{kDGLCPU, 0};
+  auto device = DeviceAPI::Get(ctx);
+  hmap_ = static_cast<Mapping*>(
```

Comment:
In general, i would prefer to have less memory ops, esp. having an exposed pointer to handle memory allocation and release. It might be error-prone and hard to debug in the future.

Consider using:
- NDArray
- Vector
- unique_ptr

Reply:
Use `unique_ptr`